include/base_loop.h

```diff
@@ -0,0 +1,213 @@
+/**
```

Comment:
can all the new files be in a separate directory? some sort of organization of the previous api vs this new set

Reply:
if we have a directory you can add a README file and add some comments of the reasoning behind this new api.

Reply:
fair point, moved to a subdir and added a readme